bundles/org.openhab.core.storage.json/src/main/java/org/openhab/core/storage/json/internal/InstantSerializer.java

```diff
@@ -0,0 +1,36 @@
+/**
+ * Copyright (c) 2010-2024 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.storage.json.internal;
+
+import java.lang.reflect.Type;
+import java.time.Instant;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+
+/**
+ * The {@link InstantSerializer} converts an {@link Instant} to a formatted UTC string.
+ *
+ * @author Jacob Laursen - Initial contribution
+ */
+@NonNullByDefault
+public class InstantSerializer implements JsonSerializer<Instant> {
+    @Override
+    public JsonElement serialize(Instant instant, Type typeOfSrc, JsonSerializationContext context) {
+        return new JsonPrimitive(instant.toString());
```

Comment:
By default this would be produced (in Java 17):

```json
"timestamp": {
  "seconds": 1704747106,
  "nanos": 550136800
}
```

Since Java 17 JDK internals are [strongly encapsulated](https://openjdk.org/jeps/403) so this is from using a reflection-based type adapter, i.e. we would be relying on implementation details outside of our control.

Reply:
Ok. I agree. But why don't we use `Instant.toEpochMilli` und `Instant.fromEpochMilli`? This is immediately backward compatible.

Reply:
Do you mean in the type adapter while still keeping `Instant` in in the DTO:
https://github.com/openhab/openhab-core/blob/6b2182dec6d77fe29c29bb563e637a31060b9fae/bundles/org.openhab.core.config.discovery/src/main/java/org/openhab/core/config/discovery/internal/DiscoveryResultImpl.java#L46

?

We could, but the only advantage IMO would be that the deserializer is a bit simpler because it doesn't need to support both old and new format. Over time, the code for converting the old format will be executed less and less since the new format is parsed first, so in terms of performance it shouldn't be a big deal.

The disadvantage would be still keeping epoch milliseconds which are less readable when inspecting the JSON file.

In both cases we need to parse a string, so the epoch version of the deserializer would still need to be:
```java
        try {
            return Instant.ofEpochMilli(Long.parseLong(content));
        } catch (NumberFormatException e) {
            throw new JsonParseException("Could not parse as Instant: " + content, e);
        }
```
(like in the fallback in the current implementation)

Reply:
Writing that made me wonder about other possibilities. so just pushed a simplification:
```java
return Instant.ofEpochMilli(element.getAsLong());
```